src/dmd/pragmasem.d

```diff
@@ -0,0 +1,472 @@
+{
+    auto psa = PragmaSemanticAnalysis(sc,ps,null);
+    psa.semantic(ps.ident);
+    return psa.result;
+}
+
+/**
+ * Semanically analyses a PragmaDeclaration.
+ * Params:
+ *  sc        = scope of evaluation
+ *  pd        = PragmaDeclaration to analyse
+ */
+extern(C++) void pragmaSemantic(Scope* sc, PragmaDeclaration pd)
+{
+    auto psa = PragmaSemanticAnalysis(sc,null,pd);
+    psa.semantic(pd.ident);
+}
+
+private struct PragmaSemanticAnalysis
+{
+    Scope* sc;
+    Expressions* args;
+    PragmaStatement ps;
+    PragmaDeclaration pd;
+    Statement result;
+
+    this(Scope* sc, PragmaStatement ps, PragmaDeclaration pd)
+    {
+        this.sc = sc;
+        this.args = ps ? ps.args : pd.args;
+        this.ps = ps;
+        this.pd = pd;
+    }
+    private void setError()
+    {
+        if (ps)
+            result = new ErrorStatement();
+    }
+    void semantic(Identifier ident)
+    {
+        import dmd.target;
+
+        if (ident == Id.msg)
+            pragmaMsgSemantic();
+        else if (ident == Id.lib)
+            pragmaLibSemantic();
+        else if (ident == Id.startaddress)
+            pragmaStartAddressSemantic();
+        else if (ident == Id.Pinline)
+            pragmaInlineSemantic();
+        else if (ident == Id.mangle)
+            return pragmaMangleSemantic(); // does semantic on its decls
+        else if (ident == Id.crt_constructor || ident == Id.crt_destructor)
+            pragmaCrtCtorDtorSemantic(ident);
+        else if (!Target.pragmaSemantic(ident,sc,args,ps,pd))
+        {
+            if (global.params.ignoreUnsupportedPragmas && global.params.verbose)
+            {
+                /* Print unrecognized pragmas
+                 */
+                OutBuffer buf;
+                buf.writestring(pd.ident.toChars());
+                bool first = true;
+                foreach (e; *args)
+                {
+                    sc = sc.startCTFE();
+                    e = e.expressionSemantic(sc);
+                    e = resolveProperties(sc, e);
+                    sc = sc.endCTFE();
+                    e = e.ctfeInterpret();
+                    if (first)
+                    {
+                        buf.writestring(" (");
+                        first = false;
+                    }
+                    else
+                        buf.writeByte(',');
+                    buf.writestring(e.toChars());
+                }
+                if (args.dim)
+                buf.writeByte(')');
+                message("pragma    %s", buf.peekString());
+            }
+            else
+            {
+                enum msg = "unrecognized `pragma(%s)`";
+                if (ps)
+                    error(ps.loc, msg, ps.ident.toChars());
+                else
+                    error(pd.loc, msg, pd.ident.toChars());
+            }
+        }
+
+        if (pd && pd.decl)
+        {
+            pd.error("is missing a terminating `;`");
```

Comment:
The provider of the pragma should determine whether this is an error or not.

It should be moved to a separate function and called by `pragmaMsgSemantic` and others.